src/main/scala/util/Annotations.scala

```diff
@@ -3,12 +3,98 @@
 package freechips.rocketchip.util
 
 import Chisel._
+import chisel3.internal.InstanceId
 import chisel3.experimental.{annotate, ChiselAnnotation, RawModule}
 import firrtl.annotations._
+import freechips.rocketchip.diplomacy._
+import freechips.rocketchip.tilelink.TLToAXI4IdMapEntry
 
-// TODO: replace this with an implicit class when @chisel unprotects dontTouchPorts
-trait DontTouch { self: RawModule =>
+/** Record a case class that was used to parameterize this target. */
+case class ParamsAnnotation(target: Named, paramsClassName: String, params: Map[String,Any]) extends SingleTargetAnnotation[Named] {
+  def duplicate(n: Named) = this.copy(n)
+}
+
+case class ParamsChiselAnnotation[T <: Product](target: InstanceId, params: T) extends ChiselAnnotation {
+  private val paramMap = params.getClass.getDeclaredFields.map(_.getName).zip(params.productIterator.to).toMap
+  def toFirrtl = ParamsAnnotation(target.toNamed, params.getClass.getName, paramMap)
+}
+
+/** Record an address map. */
+case class AddressMapAnnotation(target: Named, mapping: Seq[AddressMapEntry], label: String) extends SingleTargetAnnotation[Named] {
+  def duplicate(n: Named) = this.copy(n)
+
+  def toUVM: String =
+    s"// Instance Name: ${target.serialize}\n" +
+      mapping.map(_.range.toUVM).mkString("\n")
+
+  def toJSON: String =
+    s"""{\n  "${label}":  [\n""" +
+      mapping.map(_.range.toJSON).mkString(",\n") +
+      "\n  ]\n}"
+}
+
+/** Record a conversion of TL source ids to AXI4 ids. */
+case class TLToAXI4IdMapAnnotation(target: Named, mapping: Seq[TLToAXI4IdMapEntry]) extends SingleTargetAnnotation[Named] {
+  def duplicate(n: Named) = this.copy(n)
+}
+
+/** Marks this module as a candidate for register retiming */
+case class RetimeModuleAnnotation(target: ModuleName) extends SingleTargetAnnotation[ModuleName] {
+  def duplicate(n: ModuleName) = this.copy(n)
+}
+
+/** Annotation capturing information about port slave devices. */
+case class SlaveAddressMapChiselAnnotation(
+    target: InstanceId,
+    addresses: Seq[AddressSet],
+    perms: ResourcePermissions) extends ChiselAnnotation {
+  private val range = AddressRange.fromSets(addresses)
+  def toFirrtl = AddressMapAnnotation(
+    target = target.toNamed,
+    mapping = range.map { r => AddressMapEntry(r, perms, Nil) },
+    label = "slaves")
+}
 
+/** Record information about a top-level port of the design */
+case class TopLevelPortAnnotation(
+    target: ComponentName,
+    protocol: String,
+    tags: Seq[String],
+    mapping: Seq[AddressMapEntry]) extends SingleTargetAnnotation[ComponentName] {
+  def duplicate(n: ComponentName) = this.copy(n)
+}
+
+/** Helper object containing methods for applying annotations to targets */
+object annotated {
+  def params[T <: Product](component: InstanceId, params: T): T = {
```

Comment:
Can `Data` be annotated with `params` or just `Modules`? If it's only `Modules`, I would change this from `InstanceId` to `RawModule`

Reply:
It can also be applied to `Data` IMO. However, when I did so in practice (my example was adding the annotation to `HasBtbParameters` instead of `BTBModule`) I added a looooooooot of annotations. I'm not sure what the performance implications are of having 1000s of them, so I was more modest in practice so far.

Reply:
That's fair enough. I'm sure we'll hit some performance issues with toooooons of annotations so we should take a look at that before we have too much blow up.